crates/daphne-worker/src/aggregator/queues/async_aggregator.rs

```diff
@@ -0,0 +1,342 @@
+        part_batch_sel,
+        agg_job_id,
+        initialize_reports,
+        taskprov_advertisement,
+    }: AsyncAggregationMessage<'_>,
+) {
+    let task_id = &initialize_reports.task_id;
+    tracing::debug!("getting task config");
+    let taskprov_advertisement = match taskprov_advertisement
+        .map(|s| TaskprovAdvertisement::parse_taskprov_advertisement(&s, task_id, version))
+        .transpose()
+    {
+        Ok(taskprov_advertisement) => taskprov_advertisement,
+        Err(e) => bail!(err = e, "taskprov advertisement was illformed"),
+    };
+
+    // 1.
+    let (task_config, get_task_config_duration) = time_it(daphne::roles::resolve_task_config(
+        app,
+        &daphne::DapRequestMeta {
+            version,
+            media_type: None,
+            task_id: *task_id,
+            taskprov_advertisement,
+        },
+    ))
+    .await;
+
+    let task_config = match task_config {
+        Ok(t) => t,
+        Err(e) => bail!(retry message, err = e, "not such task config"),
+    };
+
+    let agg_param = match daphne::DapAggregationParam::get_decoded_with_param(
+        &task_config.vdaf,
+        &initialize_reports.agg_param,
+    ) {
+        Ok(param) => param,
+        Err(e) => bail!(err = e, "dap aggregation parameter was illformed"),
+    };
+
+    // 2.
+    tracing::debug!("initializing reports");
+    let (initialized_reports, initialize_reports_duration) = match time_it(
+        app.compute_offload
+            .compute::<_, compute_offload::InitializedReports>(
+                "/compute_offload/initialize_reports",
+                &initialize_reports,
+            ),
+    )
+    .await
+    {
+        (Ok(init), duration) => (init, duration),
+        (Err(e), _) => bail!(retry message, err = e, "failed to initialize reports"),
+    };
+
+    let time_precision = task_config.time_precision;
+    let state_machine = ToInitializedReportsTransition {
+        task_id: *task_id,
+        part_batch_sel,
+        task_config,
+    }
+    .with_initialized_reports(agg_param, initialized_reports.reports);
+
+    // 3.
+    tracing::debug!("checking replays");
+    let (state_machine, check_for_replays_duration) =
+        time_it(state_machine.check_for_replays(|report_ids| {
+            let report_ids = report_ids.cloned().collect::<Vec<_>>();
+            shard_reports(
+                app.durable(),
+                &initialize_reports.task_id,
+                time_precision,
+                agg_job_id,
+                report_ids.into_iter(),
+            )
+        }))
+        .await;
+
+    let state_machine = match state_machine {
+        Ok(st) => st,
+        Err(e) => bail!(retry message, err = e, "failed to check replays"),
+    };
+
+    let (span, agg_job_response) = match state_machine.finish() {
+        Ok(output) => output,
+        // this error is always caused by a bug in the code, it's not recoverable
+        Err(e) => bail!(err = e, "failed to finish aggregation"),
```

Comment:
Is a `DapAbort` possible here?

Reply:
No, `bail!` here is my own macro. See above. I renamed it to `give_up` since `bail` is already famous in `anyhow`